src/storage/src/client/controller.rs

```diff
@@ -485,14 +498,33 @@ where
                 .insert_without_overwrite(&mut self.state.stash, &id, metadata)
                 .await?;
 
-            let (write, read) = self
+            let (mut write, mut read) = self
                 .persist_client
                 .open(metadata.data_shard)
                 .await
                 .expect("invalid persist usage");
 
-            let collection_state =
-                CollectionState::new(description.clone(), read.since().clone(), metadata);
+            // Advance the collection's `since` and `upper` to the given `as_of`.
+            if let Some(as_of) = &description.as_of {
+                let lower = write.upper().clone();
+                if PartialOrder::less_than(&lower, as_of) {
+                    let empty_updates: [((SourceData, ()), T, Diff); 0] = [];
+                    write
+                        .append(empty_updates, lower, as_of.clone())
+                        .await
+                        .expect("cannot append updates")
+                        .expect("invalid/outdated upper");
+                }
+
+                read.downgrade_since(as_of.clone()).await;
+            }
+
+            let collection_state = CollectionState::new(
+                description.clone(),
+                read.since().clone(),
+                write.upper().clone(),
```

Comment:
What was the reason we passed only the `since` here previously, but not the `upper`?